drivers/adxl372/adxl372_i2c.c

```diff
@@ -0,0 +1,127 @@
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include "platform_drivers.h"
+#include "adxl372.h"
+
+/**
+ * Read from device.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_data - The register data.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t adxl372_i2c_reg_read(struct adxl372_dev *dev,
+			     uint8_t reg_addr,
+			     uint8_t *reg_data)
+{
+	uint8_t data;
+	int32_t ret;
+
+	data = reg_addr;
+
+	ret = i2c_write(dev->i2c_desc, &data, 1, 0);
+	if (ret < 0)
+		return ret;
+
+	ret = i2c_read(dev->i2c_desc, &data, 1, 0);
+	if (ret < 0)
+		return ret;
+
+	*reg_data = data;
+
+	return ret;
+}
+
+/**
+ * Write to device.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_data - The register data.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t adxl372_i2c_reg_write(struct adxl372_dev *dev,
+			      uint8_t reg_addr,
+			      uint8_t reg_data)
+{
+	uint8_t buf[2];
+
+	buf[0] = reg_addr;
+	buf[1] = reg_data & 0xFF;
+
+	return i2c_write(dev->i2c_desc, buf, ARRAY_SIZE(buf), 0);
+}
+
+/**
+ * Multibyte read from device. A register read begins with the address
+ * and autoincrements for each aditional byte in the transfer.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_data - The register data.
+ * @param count - Number of bytes to read.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t adxl372_i2c_reg_read_multiple(struct adxl372_dev *dev,
+				      uint8_t reg_addr,
+				      uint8_t *reg_data,
+				      uint16_t count)
+{
+	uint8_t buf[512];
+	uint16_t i;
+	int32_t ret;
+
+	buf[0] = reg_addr;
+	memset(&buf[1], 0x00, count);
```

Comment:
it would probably be a good idea to check that `count < sizeof(buf)` and return an err here too

Reply:
sorry;  `count < sizeof(buf) - 1`